File: Stage06-FromFile/Lua/location.lua
local Shared = require("shared")
local class = require "lib.class"
local Location = class:derive("Location")

function Location:new(name,
	             description,
	             toNorth,
	             toEast,
	             toSouth,
	             toWest,
	             items,
	             itemRequired,
	             enemy)
	self.Name 			= name
	self.Description 	= description			-- string
	self.ToNorth 		= toNorth or ""			-- string
	self.ToEast 		= toEast or ""			-- string
	self.ToSouth 		= toSouth or ""			-- string
	self.ToWest 		= toWest or ""			-- string
	self.Items 			= items or {}			-- Item keys
	self.ItemRequired 	= itemRequired or ""	-- Item key
	self.Enemy 			= enemy	or ""			-- Enemy key
end

function table.removeKey(tbl, value)
	local index = 0
	for k,v in pairs(tbl) do
		if v == value then
			index = k
			break
		end
	end
	if index > 0 then
		table.remove(tbl, index)
	end
	return tbl
end

local function trim(s)
   return (s:gsub("^%s*(.-)%s*$", "%1"))
end
string.trim = trim
local function TrimEnd(s, char)
	if s:sub(#s - #char + 1) == char then
		return s:sub(1, #s - #char)
	else
		return s
	end
	--return (s:match( "(.-)%s*$" ))
end
string.TrimEnd = TrimEnd

function Location:GetName()
	return self.Name --string
end

function Location:GetDescription()
	return self.Description --string
end

function Location:GetToNorth()
	return self.ToNorth --string
end

function Location:GetToEast()
	return self.ToEast --string
end

function Location:GetToSouth()
	return self.ToSouth --string
end

function Location:GetToWest()
	return self.ToWest --string
end

function Location:GetItems()
	return self.Items
end

function Location:GetItemRequired()
	return self.ItemRequired -- Item key
end

function Location:GetEnemy()
	return self.Enemy
end

function Location:SetEnemy(value)
	self.Enemy = value
end

--[[  public methods ]]
function Location:AddItem(item) -- add item key
	table.insert(self.Items, item)
end

function Location:RemoveItem(itemKey)
	self.Items = table.removeKey(self.Items, itemKey)
end

function Location:DisplayLocation()
	--[[  descrbe the current location, any items inside it, and exits ]]
	local exits = {}
	if self.ToNorth ~= "" then
		table.insert(exits, "north -> "..self.ToNorth) -- adds name of exit to list
	end
	if self.ToEast ~= "" then
		table.insert(exits, "east -> "..self.ToEast) -- adds name of exit to list
	end
	if self.ToSouth ~= "" then
		table.insert(exits, "south -> "..self.ToSouth) -- adds name of exit to list
	end
	if self.ToWest ~= "" then
		table.insert(exits, "west -> "..self.ToWest) -- adds name of exit to list
	end
	local row = 1
	print("You are in a ".. self.Name..", "..self.Description)
	if #exits == 0 then
		print("There are no exits")
		row = 2
	end
	if #self.Items > 0 then
		local output = "In this location there is: "
		for _, item in ipairs(self.Items) do
			output = output..item .. ", "
		end
		output = output:sub(1, output:len() - 2) -- remove comma
		print(output)
		row = row + 1
	end
	if self.Enemy ~= "" then
		print("ENEMY: "..self.Enemy.."!")
		row = row + 1
	end

	return exits, row	-- return exits list with name of exit as well as direction: {"east -> coridoor","north -> magic portal"}
end
return Location	

File: Stage06-FromFile/C#/Location.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace Adventure_06_Improvements
{
    internal class Location
    {
        public string Name { get; set; }
        public string Description { get; set; }
        public string ToNorth { get; set; }
        public string ToEast { get; set; }
        public string ToSouth { get; set; }
        public string ToWest { get; set; }
        public List<string> Items { get; set; }
        public string ItemRequired { get; set; }
        public string Enemy { get; set; }
        public Location(string name, string description, string tonorth, string toeast, string tosouth, string towest,
                                            List<string> items = null, string itemRequired = "", string enemy = "")
        {
            Name = name;
            Description = description;
            ToNorth = tonorth;
            ToEast = toeast;
            ToSouth = tosouth;
            ToWest = towest;
            Items = items ?? new List<string>();
            ItemRequired = itemRequired;
            Enemy = enemy;
        }
        public void AddItem(string item)
        {
            if(!Items.Contains(item))
                Items.Add(item);
        }
        public void RemoveItem(string item)
        {
            if(Items.Contains(item))
                Items.Remove(item);
        }
        public List<string> DisplayLocation(ref int row)
        {
            /// descrbe the current location, any items inside it, and exits ///
            List<string> exits = new List<string>();
            if(ToNorth != "")
                exits.Add($"north -> {ToNorth}");
            if(ToEast != "")
                exits.Add($"east -> {ToEast}");
            if(ToSouth != "")
                exits.Add($"south -> {ToSouth}");
            if(ToWest != "")
                exits.Add($"west -> {ToWest}");
            row = 1;
            Console.WriteLine($"You are in a {Name}, {Description}");
            if(exits.Count == 0)
            {
                Console.WriteLine("There are no exits");
                row = 2;
            }
            if(Items.Count > 0)
            {
                string output = "In this location there is: ";
                foreach(string item in Items)
                    output += item + ", ";
                output = output.Remove(output.Length - 2); // remove ', '
                Console.WriteLine(output);
                row++;
            }
            if (Enemy != "")
            {
                Console.WriteLine($"ENEMY: {Enemy}!");
                row++;
            }

            return exits; // row is by reference so does not need to be returned
        }
    }
}
